### DISCUSSION TITLE
Result of step is an update to a database

### DISCUSSION THREAD

**Discussion by stenw** [ropensci/targets#860] [Help] [answered]
In the documentation the situation is discussed in which data is extracted from a database. I have the opposite situation, a step writes a table to a postgres database. The function does not return an 'R' results so I am not sure how to handle this. I am also not sure watching the database directory for changes would work. Any ideas?

Thanks in advance,

Sten

**Comment by wlandau** [comment, marked as answer]
Is that step one of the last targets in the pipeline? In other words, when you write to the postgres database, does the target have any downstream targets that depend on it? You might consider just returning nothing, or even returning a hash of the data you push, without making `targets` track the actual database for changes. You would just need to trust that the `targets` pipeline is the only thing modifying the database. This is a pattern for continuous deployment, i.e. when a target deploys a Shiny app to shinyapps.io or RStudio Connect.

**Comment by januz** [comment]
>  You would just need to trust that the `targets` pipeline is the only thing modifying the database.

@wlandau Is this a necessary prerequisite?

I came here looking whether other people use `{targets}` with targets that write to databases and what the return value for these targets should be. The background is that I am currently thinking about how to best implement an ETL pipeline that, once or several times daily, would read from different sources (databases, APIs, files), do all kinds of transformations on the data, and store them in a destination database.

While I've also looked other tools (specific ETL tools), I would love to stay in R because it's the language most people in my team feel most comfortable with and use `{targets}` like we do in most other projects. Does anything speak against using `{targets}` in this way? Maybe because it's not necessarily in line with the original goal because most targets, potentially the complete pipeline, would be run anew every time the pipeline is executed? Another thing that might be hard is keeping a continuous log of all pipeline runs and data transfer success/failure but I guess one could solve that by having a target that collates this information and also writes it to a database.

Would love to hear your thoughts! Thanks

**Reply by wlandau** [reply]
Normally, we would expect `targets` to track the input and output files with hashes and automatically figure out which data objects changed. With databases, the data is stored in tables that `targets` does not know how to access, so it cannot detect the changes on its own. In the general case, this may cause some targets to skip when they should rerun. But if you know what you are doing and you are comfortable with how the pipeline is set up, you might still be able to use it without negative consequences.

For a target that is the furthest thing downstream, if it writes to a database, this is mostly fine. The only problem is if that target writes to a table in the database AND something else writes to the same rows in the same table of the database. `targets` won't know if any external modifications happen and thus won't rerun that part of the pipeline in response to corrupted output. But as long as you make certain that `targets` is the only thing writing to that location, you will be fine.

For an upstream target that takes an input from a database, things get trickier: the database is upstream, so it really matters that we detect changes. `targets` does not look at database tables by default, but you custom `tar_cue()`s and/or utility functions from `tarchetypes` to make that target depend on a custom hash or last-modified timestamp of the database table you are pulling from.

**Reply by januz** [reply]
Thanks so much, @wlandau! Yes, I was planning to determine the state of the upstream targets based on change sets, timestamps, etc. And sounds good for the targets that write to the destination db, I was planning for them to be the furthest thing down in the pipeline. As for making sure that no other process is writing to the same rows at the same time, that would be a concern in any setup not just with `{targets}`, right? Generally, the database should prevent corruption by locking but you're right that two processes could write conflicting data one after the other and the non-`{targets}` process could "win" without `{targets}` knowing about it. One could theoretically pull the same data `{targets}` just wrote to the database from the database and compare against the target that was written to the database?!

**Reply by wlandau** [reply]
> As for making sure that no other process is writing to the same rows at the same time, that would be a concern in any setup not just with {targets}, right? 

That is one concern. Another is if 

1. `targets` writes to the database, then 
2. some time elapses, and then
3.  someone manually writes to the same data by hand. 

Because the data is in the database, `targets` does not see (3) and will incorrectly judge the target to be up to date.

**Reply by januz** [reply]
Okay. So the target writing to the database should be set to always run I guess? That's what would be needed either way I think because it would really be moving (new) data from A to B. Thanks so much!

**Reply by wlandau** [reply]
> So the target writing to the database should be set to always run I guess? 

It depends. If you think someone else might modify the same data in the same table, e.g. by hand, then yes. But if this is not a risk at your workplace, then you probably do not need to worry about it.
